### PR TITLE
Implement matrix transpose

### LINKED ISSUE #4: Optimize matrix transposition function
Our current matrix transposition function, used during OT extension, is naively implemented to first transform the matrix into boolean vectors. This is not efficient and is a good target for optimization.

```rust
pub fn transpose(m: &Vec<Vec<u8>>) -> Vec<Vec<u8>> {
    let bits: Vec<Vec<bool>> = m.iter().map(|row| u8vec_to_boolvec(row)).collect();
    let col_count = bits[0].len();
    let row_count = bits.len();

    let mut bits_: Vec<Vec<bool>> = vec![vec![false; row_count]; col_count];
    let mut m_: Vec<Vec<u8>> = Vec::with_capacity(col_count);

    for j in 0..row_count {
        for i in 0..col_count {
            bits_[i][j] = bits[j][i];
        }
    }

    for row in bits_.iter() {
        m_.push(boolvec_to_u8vec(row));
    }

    m_
}
```

### PR DESCRIPTION
**Pull request by th4s**
Adding the matrix transpose algorithm. I open this as a draft, because it isn't yet integrated into OT. This PR resolves #4.

### CODE REVIEW

**Comment by th4s** [comment]
Adapt KOS implementation to support matrix-transpose:
- [ ]  Make sure that OT matrix has a multiple of 32 columns and rows
- [ ] Go from `<Vec<Vec<T>>` matrix representation to `<Vec<T>>`

**Review by sinui0 — APPROVED** [review]
lgtm, nice work

File: matrix-transpose/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+pub use scalar::transpose_unchecked;
+
+use thiserror::Error;
+
+/// This function transposes a matrix on the bit-level.
+///
+/// This implementation requires that the number of rows is a power of 2
+/// and that the number of columns is a multiple of 8
+pub fn transpose_bits(matrix: &mut [u8], rows: usize) -> Result<(), TransposeError> {
+    // Check that slice is rectangular
```
Comment (themighty1) [inline matrix-transpose/src/lib.rs:27]: I feel like not everyone is familiar with using & like that. Let's add a comment:
// Check that slice is rectangular, i.e. the number of cells is  a multiple of the number of rows.

**Review by themighty1 — CHANGES_REQUESTED** [review]
I ran the tests on nightly with 
cargo test --features "simd-transpose"
 and got an error

error[E0658]: trait bounds other than `Sized` on const fn parameters are unstable
   --> /home/default2/.cargo/registry/src/github.com-1ecc6299db9ec823/crossbeam-epoch-0.9.9/src/atomic.rs:314:6
    |
314 | impl<T: ?Sized + Pointable> Atomic<T> {
    |      ^
...
346 |     pub const fn null() -> Atomic<T> {
    |     -------------------------------- function declared as const here
    |
    = note: see issue #93706 <https://github.com/rust-lang/rust/issues/93706> for more information
    = help: add `#![feature(const_fn_trait_bound)]` to the crate attributes to enable

For more information about this error, try `rustc --explain E0658`.
error: could not compile `crossbeam-epoch` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed

**Comment by th4s** [comment]
Can you repeat with `cargo +nightly -F simd-transpose`? You have to compile with nightly toolchain since `simd-transpose` only works on nightly.

**Review by themighty1 — APPROVED** [review]
well done, ack.

**Pull request merged**